serde_json_path_core/src/spec/selector/filter.rs

```diff
@@ -0,0 +1,602 @@
+//! Types representing filter selectors in JSONPath
+use serde_json::{Number, Value};
+
+use crate::spec::{
+    functions::{FunctionExpr, JsonPathValue, Validated},
+    query::{Query, QueryKind, Queryable},
+    segment::{QuerySegment, Segment},
+};
+
+use super::{index::Index, name::Name, Selector};
+
+mod sealed {
+    use serde_json::Value;
+
+    use crate::spec::functions::FunctionExpr;
+
+    use super::{BasicExpr, ComparisonExpr, ExistExpr, LogicalAndExpr, LogicalOrExpr};
+
+    pub trait Sealed {}
+    impl Sealed for Value {}
+    impl Sealed for LogicalOrExpr {}
+    impl Sealed for LogicalAndExpr {}
+    impl Sealed for BasicExpr {}
+    impl Sealed for ExistExpr {}
+    impl Sealed for ComparisonExpr {}
+    impl<V> Sealed for FunctionExpr<V> {}
+}
+
+/// Trait for testing a filter type
+pub trait TestFilter: sealed::Sealed {
+    /// Test self using the current and root nodes
+    fn test_filter<'b>(&self, current: &'b Value, root: &'b Value) -> bool;
+}
+
+impl TestFilter for Value {
+    fn test_filter<'b>(&self, _current: &'b Value, _root: &'b Value) -> bool {
+        match self {
+            Value::Null => false,
+            Value::Bool(b) => *b,
+            Value::Number(n) => n != &Number::from(0),
+            _ => true,
```

Comment:
This should be unreachable?

Reply:
This seems an impl defined manner. RFC 9535 doesn't define coere between any value to LogicalType value.